pss/pss_test.go

```diff
@@ -507,116 +510,128 @@ func TestMessageProcessing(t *testing.T) {
+	// The message will be retried once we send to continueC, so first, we change the forward function
+	ps.forwardFunc = forward
+	continueC <- struct{}{}
+	<-successC
+	if len(processed) != 2 {
+		t.Fatal("There should be two message processed", "len(processed)", len(processed))
 	}
 
-	// full address mismatch should put message in queue
-	msg.To[0] = 0xff
-	if err := ps.handle(context.TODO(), msg); err != nil {
+}
+
+func TestOutboxFull(t *testing.T) {
+	// setup
+	privkey, err := crypto.GenerateKey()
+	if err != nil {
 		t.Fatal(err.Error())
 	}
-	tmr.Reset(time.Millisecond * 10)
-	outmsg = nil
-	select {
-	case outmsg = <-ps.outbox:
-	case <-tmr.C:
-	}
-	if outmsg == nil {
-		t.Fatal("expected message in outbox on address mismatch, but empty")
-	}
-	outmsg = nil
-	select {
-	case outmsg = <-ps.outbox:
-	default:
-	}
-	if outmsg != nil {
-		t.Fatalf("expected only one queued message but also had message %v", msg)
+
+	addr := make([]byte, 32)
+	addr[0] = 0x01
+	ps := newTestPssStart(privkey, network.NewKademlia(addr, network.NewKadParams()), NewParams(), false)
+	defer ps.Stop()
+	outboxCapacity := 2
+
+	processed := make([]*PssMsg, 0)
+	procChan := make(chan struct{})
+	ps.outbox = newOutbox(outboxCapacity)
+	succesForward := func(msg *PssMsg) bool {
+		log.Info("Processing message")
+		<-procChan
+		log.Info("Message processed")
+		processed = append(processed, msg)
+		return true
 	}
+	ps.forwardFunc = succesForward
+
+	ps.Start(nil)
+}
 
-	// expired message should be dropped
-	msg.Expire = uint32(time.Now().Add(-time.Second).Unix())
-	if err := ps.handle(context.TODO(), msg); err != nil {
+// Stopping ps in the middle of a message process should not produce an error trying to write in a closed channel
+func TestMessageOutboxClose(t *testing.T) {
+	// setup
+	privkey, err := crypto.GenerateKey()
+	if err != nil {
 		t.Fatal(err.Error())
 	}
-	tmr.Reset(time.Millisecond * 10)
-	outmsg = nil
-	select {
-	case outmsg = <-ps.outbox:
-	case <-tmr.C:
-	default:
-	}
-	if outmsg != nil {
-		t.Fatalf("expected empty queue but have message %v", msg)
-	}
 
-	// invalid message should return error
-	fckedupmsg := &struct {
-		pssMsg *PssMsg
-	}{
-		pssMsg: &PssMsg{},
-	}
-	if err := ps.handle(context.TODO(), fckedupmsg); err == nil {
-		t.Fatalf("expected error from processMsg but error nil")
+	addr := make([]byte, 32)
+	addr[0] = 0x01
+	ps := newTestPssStart(privkey, network.NewKademlia(addr, network.NewKadParams()), NewParams(), false)
+
+	procChan := make(chan struct{})
+	succesForward := func(msg *PssMsg) bool {
+		log.Info("Processing message")
+		<-procChan
+		log.Info("Message processed")
+		return true
 	}
+	ps.forwardFunc = succesForward
 
-	// outbox full should return error
-	msg.Expire = uint32(time.Now().Add(time.Second * 60).Unix())
+	ps.Start(nil)
+	err = ps.enqueue(testRandomMessage())
+
+	// stopping before processing message
+	ps.Stop()
+
+	// finish processing message
+	procChan <- struct{}{}
```

Comment:
Shouldn't we here instead just check if the channel is closed if that's what the test is about? The error state of a test should not be `panic`

```
select {
    case _, ok := <-procChan
        if !ok {
            t.Fatal(...)
        }
     default:
}
```

Reply:
The idea was closing ps /outbox in the middle of a message processing and avoid panics (because of closed channels). Not sure how to test that.

Reply:
Which channel did you want to test for panic on close?

Reply:
Well, `process` and `slots` channels because they are the ones that could be used by one routine processing a message (`process` if the forwarding failed in reenqueue(), `slots` if succeded in free())

Reply:
I'm not sure this should be tested. But if the problem is the attempt to re-enqueue on fail after channel is closed then maybe the re-enqueue should have a check on the quit channel instead on higher priority, and if its closed then don't even try to re-enqueue. Or am I missing the point?

Reply:
Maybe we shouldn't test that, but in the early stages of implementation I had `panic`'s when closing so I did this test to feel safe about it.
I can remove it completely.

Reply:
Test removed